BJS2-68727

tanufa25 opened the pull request:
Vesteross library logic is added

Commit:
BJS2-68727
Vesteross library logic is added

iBerbasov commented on src/main/java/school/faang/bjs2_68727/vesteross_library/Book.java:
```diff
@@ -0,0 +1,16 @@
+    private String title;
+    private String author;
+    private int year;
+
```
Лишние пустые строки

iBerbasov commented on src/main/java/school/faang/bjs2_68727/vesteross_library/LibrarySystem.java:
```diff
@@ -0,0 +1,37 @@
+    Map<Book, String> libraryMap = new HashMap<>();
+
+    public void addBook(String title, String author, int year, String location) {
+        libraryMap.putIfAbsent(new Book(title, author, year), location);
```
А зачем здесь putIfAbsent ? Разве обычного put не будет достаточно?

iBerbasov commented on src/main/java/school/faang/bjs2_68727/vesteross_library/Main.java:
```diff
@@ -0,0 +1,29 @@
+        librarySystem.removeBook("Kolobok", "public", 1850);
+        librarySystem.removeBook("Teremok", "public", 1852);
+        librarySystem.printAllBooks();
+
```
Лишняя пустая строка

iBerbasov commented on src/main/java/school/faang/bjs2_68727/vesteross_library/Main.java:
```diff
@@ -0,0 +1,29 @@
+        librarySystem.printAllBooks();
+
+    }
+
```
Лишние пустые строки

iBerbasov commented on src/main/java/school/faang/bjs2_68727/vesteross_library/Main.java:
```diff
@@ -0,0 +1,29 @@
+@Slf4j
+public class Main {
+    public static void main(String[] args) {
+        LibrarySystem librarySystem = new LibrarySystem();
```
А вот в main можно добавить пустых строк. Как раз в этом случае они не помешали бы - разделять логические блоки длинного метода.

Пусть здесь и в упрощенном виде без специализированных фреймворков, но тем не менее - это тесты твоей программы.
Обычно подобные unit тесты оформляют в стиле ААА - arrange act assert
Почитай об этом, раздели пустыми строками подготовку тестовых данных и наполнение библиотеки, вызов основных методов

iBerbasov commented on src/main/java/school/faang/bjs2_68727/vesteross_library/LibrarySystem.java:
```diff
@@ -0,0 +1,37 @@
+        if (location == null) {
+            log.error(BOOK_NOT_FOUND, title, author, year);
+        }
+        return location;
```
В случае , если книга не найдется - вернется null, а это очень плохая практика. Подумай, как можно это исправить.
Может лучше возвращать строку с сообщением BOOK_NOT_FOUND или вовсе выбрасывать исключение
Так же и в методе removeBook

iBerbasov replied:
Вижу, что был выбран вариант с исключением. Мне он тоже больше нравится )
Задача зачтена, но единственная рекомендация на будущее - выбрасывать RuntimeException это тоже не очень хорошая практика. Лучше выбирать более близкое к текущей ситуации (NoSuchElementException например), а еще лучше создавать свои исключения, например здесь отлично подошло бы BookNotFoundException.
C кастомным исключением и метод main станет гораздо более читабельным - сразу видно, что может пойти не так
```
try {
            String location = librarySystem.findBook("Teremok", "public", 1852);
            log.info(BOOK_LOCATION_INFO, "Teremok", location);
        } catch (BookNotFoundException e) {
            log.error(e.getMessage());
        }
}
```
В дополнение - старайся выбрасывать исключения с более информативными сообщениями. Это сильно облегчат работу потом, если сразу в логах видно не только, что книга не найдена, но и параметры, по которым она не нашлась

iBerbasov commented on src/main/java/school/faang/bjs2_68727/vesteross_library/LibrarySystem.java:
```diff
@@ -0,0 +1,37 @@
+
+    public void removeBook(String title, String author, int year) {
+        this.findBook(title, author, year);
+        libraryMap.remove(new Book(title, author, year));
```
При обращении к полям инстанса хорошей практикой считается использование this
А вот строкой выше при вызовах методов, наоборот :) this обычно опускают

Commit:
BJS2-68727
Some empty lines are removed.
Some empty lines are added for "arrange act assert" style.
Put is used instead of putIfAbsent.
RuntimeException is added when book is not found.
This for instance is added, this of method is removed.
